{% extends 'base.html' %}

{% block body %}
  <script>
  clientUniqueID=Math.random().toString(36).slice(-5);
  clientUniqueID='12345';//for one user open multi-browsers
  flag_open=0;
  integrated=0;
  label_selected_value='';
  introFlag=1;
  tourCompleted=false;
  $('#next2EDA').attr('disabled', true);
  function startTour(){
	introFlag=1;
	$('#tabs').tabs('option', 'active', 0);
	intro = introJs().setOptions({
		steps:[{
			title:'Welcome',
			intro:"<div style='width:500px'> Welcome to RNAcare, {{ user.get_username }}</div>"
		},
		{
			element: document.querySelector('.first'),
			intro:'You can upload your expression data files and meta data file here. You can skip this step and just use built-in ones in the next tab for trial.'
		},
		{
			element: document.querySelector('#tooltip1'),
			intro:'Anytime you have a question, just hover on it!'
		},
		{
			element: document.querySelector('#upload1'),
			intro:"Upload csv (bulked) RNA-seq Files, each file needs to include ID_REF",
		},
		{
			element: document.querySelector('#upload2'),
			intro:"Upload a csv meta-data File, each file needs to include ID_REF and LABEL, may include some clinic data(numeric)"
		},
		{
			element: document.querySelector('#TemplateUpload'),
			intro:'You can download the templates for upload here.'
		},
		{
			element: document.querySelector('.second'),
			intro:'You can combine different datasets including the selection of built-in ones, and define the data transformation and integration. For data transformation, you can pick log2, and feature reduction methods, like t-SNE, UMAP and PCA; For integration, the user can choose between Harmony and Combat. <b>Besides</b> It supports user\'s self-defined labels based on the clinical parameters after your integration. '
		},
		{
			element: document.querySelector('#built-in-data'),
			intro:'You can choose some of the built-in data here alone, or integrated with your previous uploaded data. Columns sharing the same column names will be retained.'
		},
		{
			element: document.querySelector('#processOption'),
			intro:'Choose options to process the data.'
		},
		{
			element: document.querySelector('#next2'),
			intro:'Click \'Processing\' button to process the data first.'
		},
		{
			element: document.querySelector('#next2EDA'),
			intro:'After processing the data, then click this to navigate to the next page.'
		},
		{
			element: document.querySelector('#metaAdvancedSettingClick'),
			intro:'Or, before navigating, you can create new labels based on the processed clinical data. In you uploaded clinical data, features with continous values can be used to create new labels here, features with categorical values will show as candidate labels directly, waiting for being activated.'
		},
		{
			element: document.querySelector('.third'),
			intro:'You can visualise the quality of the transformation and integration and download the processed data here'
		},
		{
			element: document.querySelector('#VbOption'),
			intro:'Choose the Value for coloring the processed data.'
		},
		{
			element: document.querySelector('.fourth'),
			intro:'Clustering based on you processed data and finding the biomarkers for each cluster.'
		},
		{
			element: document.querySelector('.fifth'),
			intro:'You can visualise the Violin, density and heatmap plots for your specific labels/clusters.'
		},
		{
			element: document.querySelector('.sixth'),
			intro:'Lasso regression for your specific label/cluster to find its related genes of interest.'
		},
		{
			element: document.querySelector('.seventh'),
			intro:'Enrichment analysis for your label/cluster if you don\'t upload clinical parameters in the meta data file.'
		},
		{
			element: document.querySelector('#tour'),
			intro:'Or go back to this tour again.'
		},
	],
	});
	
	intro.onchange(function(targetElement) {
      if ($(targetElement).closest('.second').length) {
        $('#tabs').tabs('option', 'active', 1); // Activate Tab 2
      }
	  else if ($(targetElement).closest('.third').length) {
        $('#tabs').tabs('option', 'active', 2); // Activate Tab 3
      }
	  else if ($(targetElement).closest('.fourth').length) {
        $('#tabs').tabs('option', 'active', 3); // Activate Tab 4
      }
	  else if ($(targetElement).closest('.fifth').length) {
        $('#tabs').tabs('option', 'active', 4); // Activate Tab 5
      }
	  else if ($(targetElement).closest('.sixth').length) {
        $('#tabs').tabs('option', 'active', 5); // Activate Tab 6
      }
	  else if ($(targetElement).closest('.seventh').length) {
        $('#tabs').tabs('option', 'active', 6); // Activate Tab 7
      }
	  else if ($(targetElement).closest('#tour').length) {
        $('#tabs').tabs('option', 'active', 0); // Activate Tab 1
      }
    });

	intro.oncomplete(function() {
    	tourCompleted = true; // Mark as completed
	});

	intro.onexit(function() {
		if (!tourCompleted) {
			introFlag=0;
    	}
	});

	intro.start();
  }
  function Tour2(){
	if(introFlag==0)return;
	intro = introJs().setOptions({
		steps:[{
			element: document.querySelector('#VbOption'),
			intro:'Choose the Value for coloring the processed data.'
		},
		{
			element: document.querySelector('#VbL'),
			intro:'Visualization by the chosen Label.'
		},
		{
			element: document.querySelector('#VbF'),
			intro:'Visualization by FileNames(batches).'
		},
		{
			element: document.querySelector('#graphDiv21A'),
			intro:'Download the image.'
		},
	],
	});
	intro.start();
  }
  function Tour3(){
	if(introFlag==0)return;
	intro = introJs().setOptions({
		steps:[{
			element: document.querySelector('#VbC'),
			intro:'Visualization by Clusters'
		},
		{
			element: document.querySelector('#DE4C'),
			intro:'Dotplot for top 4 genes of each cluster.'
		},
	],
	});
	intro.start();
  }
  document.addEventListener('DOMContentLoaded',function(){
	startTour();
  });
  $( function() {
	initDf();
    $(document).tooltip();
	$('#metaAdvancedSetting').fadeOut();
    $('#FR_processed').prop('checked',false);
	$('#my-select-label').multiSelect({ 
		keepOrder: true,selectableHeader:'<div class="custom-header">Select for coloring</div>',
		selectionHeader:'<div class="custom-header">Established Labels</div>'
	});
    tabs=$('#tabs');
	tabs.tabs();
	tabs_label=$('#tabs-label');
	tabs_label.tabs();
	$('#tabs-eda').tabs({
		activate: function (event, ui) {
			const tabId = ui.newPanel.attr("id");
			if (tabId === "tabs-eda-2") {
				$('#graphDiv11 .img1').each(function () {
					// Ensure the image is visible and rendered
					if (!$(this).data('wheelzoom-applied')) {
							wheelzoom(this);
							$(this).data('wheelzoom-applied', true);
					}
				});
			}
		}
  	});
	$('#tabs-dgea').tabs({
		activate: function (event, ui) {
			const tabId = ui.newPanel.attr("id");
			if (tabId === "tabs-dgea-2") {
				$('#graphDiv1-cluster-gene .img2').each(function () {
					// Ensure the image is visible and rendered
					if (!$(this).data('wheelzoom-applied')) {
							wheelzoom(this);
							$(this).data('wheelzoom-applied', true);
					}
				});
			}
		}
  	});
    $('#HPara').hide();
    $('#KPara').hide();
    $('#tab3div').hide();$('#tab4div').hide();
    $('#my-select').multiSelect({ keepOrder: true });
    $('.downloadData1').hide();
	$('#menu_lasso').menu();

    $('#upload1').on('click',function(){
		form_data=new FormData();
		ins=document.getElementById('multiFiles').files.length;
		if(ins==0){
			$('#msg').html('<div class="alert-danger" role="alert">Select at least one file for Gene Expression</div>');
			return;
		}
		else{
		$('#msg').html('');
		}
		
		for(var i=0;i<ins;i++){
			form_data.append("files[]", document.getElementById('multiFiles').files[i]);
		}
		form_data.append('cID', clientUniqueID);
		csrf_token=$('input[name="csrfmiddlewaretoken"]').val();
		form_data.append("csrfmiddlewaretoken",csrf_token);
		
		$.ajax({
			url: '/geneExpression/?cID='+clientUniqueID,
			dataType:'text',
			cache:false,
			contentType:false,
			processData:false,
			data:form_data,
			type:'post',
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('#preview1').html(response);
				$('#preview2').html(' ');
				
			},
			error:function(response){
				alert(response.responseText);
			}
		});
	});
	
   $('#upload2').on('click',function(){
		form_data=new FormData();
		ins=document.getElementById('singleFile').files.length;
		if(ins==0){
			$('#msg').html('<div class="alert-danger" role="alert">Select at least one file for Meta-data</div>');
			return;
		}
		else{
		$('#msg').html('');
		}
		
		form_data.append("meta", document.getElementById('singleFile').files[0]);
		csrf_token=$('input[name="csrfmiddlewaretoken"]').val();
		form_data.append("csrfmiddlewaretoken",csrf_token);
		form_data.append('cID', clientUniqueID);
		
		$.ajax({
			url: '/meta/?cID='+clientUniqueID,
			dataType:'text',
			cache:false,
			contentType:false,
			processData:false,
			data:form_data,
			type:'post',
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('#preview2').html(response);
				
			},
			error:function(response){
				alert(response.responseText);
			}
		});
	});
	
   $('#slider1').slider({
		min:0,
		max:2,
		step:0.001,
		value:1,
		slide:function(event,ui){
			$('#sliderValue').val(ui.value);
		}
	});
	$('#sliderValue').val($('#slider1').slider('value'));
	$('#sliderValue').on('change',function(){
		$('#slider1').slider('value',$(this).val());
		$('#slider1').slider('refresh');
	});
	
   $('#next1').on('click',function(){	
		tabs.tabs('option','active',1);
	});
   $('#next2').on('click',function(){
   	$('.btn').attr('disabled',true);
	url='/eda/integrate/?correct='+$('input[name="Comb"]:checked').val()+'&log2='+$('input[name="Log"]:checked').val()+'&fr='+$('input[name="FR"]:checked').val()+'&integrate='+$('#my-select').val();
	url+='&cID='+clientUniqueID;
		$.ajax({
			url: url,
			dataType:'text',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
				$('#loader').hide();
			},
			success:function(response){
				$('.btn').attr('disabled',false);
				alert('Data Process successful.');
				$('#my-select-label-numeric').html("<option value='' selected='selected'>----SELECT----</option>");
				$('#my-select-label-delete').html("<option value='' selected='selected'>----SELECT----</option>");
				$('#label-numeric-input').val('');
				$('#label-numeric-input-texts').val('');
				$('#eda-label-choose').val("<option value='batch2' selected='selected'>LABEL</option>");
				$('#next2EDA').attr('disabled', false);

				$('#tab3div').hide();
				$('#tab4div').hide();
				$('#targetGene').val('');
				$('#candidateCluster').val('');
				$('#candidateGeneList').val('');
				$('#vlnPlot').html('');
				$('#densityPlot').html('');
				$('#heatmapPlot').html('');
				$('#lassoPic').html('');
				$('#goEnrichPic').html('');
				$('#lasso_label_name').html("<option value='' selected='selected'>----SELECT----</option>");
				$('#cluster_num1').html("<option value='' selected='selected'>----SELECT----</option>");
				$('#go_label_name').html("<option value='' selected='selected'>----SELECT----</option>");
				$('#cluster_num').html("<option value='' selected='selected'>----SELECT----</option>");
				integrated=1;
				if(flag_open==1){
					metaAdvanced();
				}
			},
			error:function(response){
				alert(response.responseText);$('#loader').hide();
				$('.btn').attr('disabled',false);
				return;
			}
		});
		$('.btn').attr('disabled',false);
	});

   $('#next2EDA').on('click',function(){
	orig_label=$('#eda-label-choose').val();
	url='/eda/?cID='+clientUniqueID;
	url+='&label='+$('#eda-label-choose').val();

	$('#candidateCluster').html('<option value="" selected="selected"></option><option value="cluster">Clustering Method</option>');
	$.ajax({
			url: url,
			dataType:'json',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			},
			success:function(response){
				$('#tab3div').show();
				tabs.tabs('option','active',2);
				dfs1 = response.dfs1;
				plotlyPlot('graphDiv1',dfs1,response.method);
				$('.btn').attr('disabled',false);
				dfs2 = response.dfs2;
				plotlyPlot('graphDiv2',dfs2,response.method);
				$('#eda-label-choose').html('');
				
				for(i=0;i<response.labels.length;i++)
				{
					temp_r=response.labels[i];
					if(temp_r=='LABEL')temp_r='batch2';
					if(temp_r==orig_label){
						$('#eda-label-choose').append("<option value='"+temp_r+"' selected='selected'>"+response.labels[i]+"</option>");
						$('#candidateCluster').append("<option value='"+temp_r+"'>"+response.labels[i]+"</option>");
					}
					else{
						$('#eda-label-choose').append("<option value='"+temp_r+"'>"+response.labels[i]+"</option>");
						$('#candidateCluster').append("<option value='"+temp_r+"'>"+response.labels[i]+"</option>");
					}
				}	
				$.ajax({
					url: '/dgea/?cID='+clientUniqueID+'&label='+$('#eda-label-choose').val(),
					dataType:'json',
					cache:false,
					contentType:false,
					processData:false,
					type:'get',
					async:true,
					complete:function(){
						$('#loader').hide();
					},
					success:function(response){
						$('#graphDiv11A').attr('href','data:image/svg+xml;base64,'+response[0]);
						$('#graphDiv21A').attr('href','data:image/svg+xml;base64,'+response[1]);
						$('#graphDiv11').html('<img class="img1" src="data:image/svg+xml;base64,'+response[0]+'" width="auto" height="400px"/>');
						$('#graphDiv21').html('<img class="img1" src="data:image/svg+xml;base64,'+response[1]+'" width="auto" height="400px"/>');
						alert('Calculation Succeed!');
						
						$('#tabs-eda').tabs('option', 'active', 0)
						$('#graphDiv21 .img1').each(function () {
							wheelzoom(this);
						});
						Tour2();				
					},
					error:function(response){
						alert(response.responseText);$('#loader').hide();
						$('.btn').attr('disabled',false);
						$('#graphDiv11').html('');
						$('#graphDiv21').html('');
						$('#graphDiv11A').attr('href','...');
						$('#graphDiv21A').attr('href','...');
						return;
					}
				});
	
			},
			error:function(response){
				alert(response.responseText);$('#loader').hide();
				$('.btn').attr('disabled',false);
				return;
			}
		});
		//lasso_label_name
		url='/meta/columns/?numeric=0';
		url+='&cID='+clientUniqueID;
		$('#lasso_label_name').html("<option value='' selected='selected'>----SELECT----</option>");
		$('#cluster_num1').html("<option value='' selected='selected'>----SELECT----</option>");
		$('#go_label_name').html("<option value='' selected='selected'>----SELECT----</option>");
		$('#cluster_num').html("<option value='' selected='selected'>----SELECT----</option>");
		$.ajax({
			url: url,
			dataType:'json',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			success:function(response){
				for(i=0;i<response.length;i++){
					if(response[i][1]=='1'){
						if(response[i][0]!='LABEL'){
							$('#lasso_label_name').append('<option value="'+response[i][0]+'">'+response[i][0]+'</option>');
							$('#go_label_name').append('<option value="'+response[i][0]+'">'+response[i][0]+'</option>');
						}
						else{
							$('#lasso_label_name').append('<option value="batch2">LABEL</option>');
							$('#go_label_name').append('<option value="batch2">LABEL</option>');
						}
					}
				}
			},
			error:function(response){
				return;
			}
		});

		$('.btn').attr('disabled',false);
	});

   $('#next3').on('click',function(){
		tabs.tabs('option','active',3);
		$('#tab4div').hide();
	});
   $('#next4').on('click',function(){
		tabs.tabs('option','active',4);
	});

   $("input[type=radio][name=clustering]").on('change',function () {
            if($(this).val()=='LEIDEN')
            {
            	$('#HPara').hide();
            	$('#KPara').hide();
            	$('#LPara').show();
            }
            else if($(this).val()=='HDBSCAN'){
            	$('#LPara').hide();
            	$('#KPara').hide();
            	$('#HPara').show();
            }
            else{
            	$('#LPara').hide();
            	$('#HPara').hide();
            	$('#KPara').show();
            }
    });

  });

function initDf(){
	$('#preview1').html('');
	$('#preview2').html('');
	$.ajax({
			url: '/geneExpression/?cID='+clientUniqueID,
			dataType:'text',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('#preview1').html(response);
			},
			error:function(response){
				alert(response.responseText);
			}
	});

	$.ajax({
			url: '/meta/?cID='+clientUniqueID,
			dataType:'text',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('#preview2').html(response);
				
			},
			error:function(response){
				alert(response.responseText);
			}
		});
}
  
  
function unpack(rows,n){
	return rows.map(function(row)
	{return row[n];});}
	
function plotlyPlot(idP,dfsP,method){
	data=[];
	sequentialScale=d3.scaleSequential().domain([0,Object.keys(dfsP).length]).interpolator(d3.interpolateRainbow);
	i=0;
	for(const [key,value] of Object.entries(dfsP)){
		var trace={
		x:unpack(value.data, 0), y: unpack(value.data, 1),
		mode: 'markers',
		marker: {
			size: 10,color:sequentialScale(i),
			symbol:'circle',
			opacity: 0.5},
		type: 'scatter',
		name: key,
		text: value.obs,
		};
		data.push(trace);
		i++;
	}
	var layout = {
		legend:{
	  	x:1,
	  	y:0.5,
		itemsizing: 'constant',
	  	},
		  xaxis: {
            title: method + '1', // Dynamic x-axis name
        },
        yaxis: {
            title: method + '2', // Dynamic y-axis name
        },
	  };
	var config={
		toImageButtonOptions:{
			format:'svg',
			filename:'download',
			scale:'1'
		}
	};
	Plotly.newPlot(idP, data, layout, config);
}

function plotlyPlot1(idP,dfsP){
	data=[];
	sequentialScale=d3.scaleSequential().domain([0,Object.keys(dfsP).length]).interpolator(d3.interpolateRainbow);
	i=0;
	for(const [key,value] of Object.entries(dfsP)){
		var trace={x:value.x,y:value.y,name:value.name,type:value.type,marker:{color:sequentialScale(i)}};
		data.push(trace);
		i++;
	}
	layout={barmode:'stack'};
	var config={
		toImageButtonOptions:{
			format:'svg',
			filename:'download',
			scale:'1'
		}
	};
	Plotly.newPlot(idP, data, layout, config);
}

function clusterBtnClick(){
	if($('input[type=radio][name=clustering]:checked').val()=='LEIDEN')
        	url='/cluster/?cluster=LEIDEN&param='+$('#sliderValue').val();
    else if($('input[type=radio][name=clustering]:checked').val()=='HDBSCAN')
        	url='/cluster/?cluster=HDBSCAN&param='+$('#hdbP').val();
    else
        	url='/cluster/?cluster=Kmeans&param='+$('#kP').val();

		url+='&cID='+clientUniqueID;
        data=[];
        $('.btn').attr('disabled',true);
        $.ajax({
			url: url,
			dataType:'json',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('#tab4div').show();
				plotlyPlot('graphDiv1-cluster',response.traces, response.method);
				plotlyPlot1('bc1',response.bc1);
				plotlyPlot1('bc2',response.bc2);
				$('#graphDiv1-cluster-gene-A').attr('href','data:image/svg+xml;base64,'+response.fileName);
				$('#graphDiv1-cluster-gene').html('<img class="img2" src="data:image/svg+xml;base64,'+response.fileName+'" width="auto" height="400"/>');
				$('#graphDiv1-cluster-top-gene-A').attr('href','data:image/svg+xml;base64,'+response.fileName1);
				$('#graphDiv1-cluster-top-gene').html('<img class="img2" src="data:image/svg+xml;base64,'+response.fileName1+'" width="1000" height="auto"/>');
				alert('Calculation Succeed!');

				$('#tabs-dgea').tabs('option', 'active', 0);
				$('#graphDiv1-cluster-top-gene .img2').each(function () {
                    wheelzoom(this);
                });


				$('.downloadData1').show();
				$('#cluster_num').html("<option value='' selected='selected'>----SELECT----</option>");
				$('#cluster_num1').html("<option value='' selected='selected'>----SELECT----</option>");
				$('.btn').attr('disabled',false);
				$('#targetGene').val('')
				$('#candidateGeneList').val('');
				$('#vlnPlot').html('');
				$('#densityPlot').html('');
				$('#heatmapPlot').html('');

				if(!$('#lasso_label_name').html().includes('Cluster'))
				{
					$('#lasso_label_name').append('<option value="cluster">Cluster</option>');
					$('#go_label_name').append('<option value="cluster">Cluster</option>');
				}
				Tour3();
			},
			error:function(response){
				alert(response.responseText);
				$('#graphDiv1-cluster-gene').html('');
				$('#graphDiv1-cluster-top-gene').html('');
				$('#graphDiv1-cluster-gene-A').attr('href','...');
				$('#graphDiv1-cluster-top-gene-A').attr('href','...');
				$('#cluster_num').html("<option value='' selected='selected'>----SELECT----</option>");
				$('#cluster_num1').html("<option value='' selected='selected'>----SELECT----</option>");
				$('.btn').attr('disabled',false);
				return;
			}
		});
	
}

function goEnrich(){
	if($('#go_label_name').val()==''||$('#cluster_num').val()==''){alert('Please Select a value.');return;}
	$('.btn').attr('disabled',true);
	url='/goenrich/?colName='+$('#go_label_name').val()+'&cluster_n='+$('#cluster_num').val();
	url+='&cID='+clientUniqueID;
	$.ajax({
			url:url,
			dataType:'text',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('#goEnrichPic').html('<img class="imgGo" src="data:image/svg+xml;base64,'+response+'" width="500" height="auto"/><br><a href="data:image/svg+xml;base64,'+response+'" download="image.svg" target="_blank">Download Image</a>');
				alert('Calculation Succeed!');
				wheelzoom(document.querySelectorAll('img.imgGo'));
				$('.btn').attr('disabled',false);
			},
			error:function(response){
				alert(response.responseText);
				$('#goEnrichPic').html('');
				$('.btn').attr('disabled',false);
				return;
			}
	});
}

function golasso(){
	if($('#lasso_label_name').val()==''||$('#cluster_num1').val()==''){alert('Please Select a value.');return;}
	$('.btn').attr('disabled',true);
	url='/lasso/?colName='+$('#lasso_label_name').val()+'&cluster_n='+$('#cluster_num1').val();
	url+='&cID='+clientUniqueID;
	$.ajax({
			url:url,
			dataType:'text',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('#lassoPic').html('<img class="imgLasso" src="data:image/svg+xml;base64,'+response+'" width="1000" height="auto"></img><br><a href="data:image/svg+xml;base64,'+response+'" download="image.svg" target="_blank">Download Image</a>');
				alert('Calculation Succeed!');
				wheelzoom(document.querySelectorAll('img .imgLasso'));
				$('.btn').attr('disabled',false);
			},
			error:function(response){
				alert(response.responseText);
				$('#lassoPic').html('');
				$('.btn').attr('disabled',false);
				return;
			}
	});
}

function getTopGenes(t){
	url='';
	if(t==1){
		url='/dgea/?clusters=fileName&topN=0';
	}
	else if(t==2){
		url='/dgea/?clusters=label&topN=0';
	}
	else if(t==3){
		url='/dgea/?clusters='+$('input[type=radio][name=clustering]:checked').val()+'&topN=0';//LEIDEN,HDBSCAN,Kmeans
	}
	else if(t==4){
		url='/processedFile/corrected/?usr={{ user.get_username }}';
	}
	else if(t==5){
		url='/processedFile/correctedCluster/?usr={{ user.get_username }}';
	}
	url+='&label='+$('#eda-label-choose').val();
	url+='&cID='+clientUniqueID;
	$('.btn').attr('disabled',true);
	$.ajax({
			url:url,
			dataType:'text',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('.btn').attr('disabled',false);
				const blob = new Blob([response], { type: 'text/csv' });
   				 // Create a link element to trigger the download
				const link = document.createElement('a');
				link.href = window.URL.createObjectURL(blob);
				link.download = 'export.csv';

				// Append the link to the document and trigger a click event
				document.body.appendChild(link);
				link.click();

				// Remove the link from the document
				document.body.removeChild(link);
			},
			error:function(response){
				alert(response.responseText);
				$('.btn').attr('disabled',false);
				return;
			}
	});
}

function getSep()
{
	sep=$('#separator').val();
	if(sep=='Space')sep=' ';
	else if(sep=='Return')sep='\n';
	else if(sep=='Comma')sep=',';
	else if(sep=='Tab')sep='\t';
	else sep=';';
	return sep;
}

function loadCandiGenes()
{
	sep=getSep();
	tempValue=$('#candidateGeneList').val();
	tempCluster=$('#candidateCluster').val();
	if(tempValue==''||tempCluster=='')return;
	if(tempValue!='pca')tempValue=tempCluster;
	url='/dgea/candiGenes/?method='+tempValue;
	url+='&cID='+clientUniqueID;
	$('.btn').attr('disabled',true);
	$.ajax({
			url:url,
			dataType:'json',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('.btn').attr('disabled',false);
				$('#targetGene').val(response.join(sep));
			},
			error:function(response){
				alert(response.responseText);
				$('.btn').attr('disabled',false);
				return;
			}
	});
}

function TGenePlotClick()
{
	tempValue=$('#targetGene').val();
	groupby=$('#candidateCluster').val();
	if(groupby=='')return;
	sep=getSep();
	tempValue=tempValue.split(sep).join(',');
	if(tempValue=='')return;
	url='/dgea/plot/?type=vln&geneList='+tempValue+'&groupby='+groupby;
	url+='&cID='+clientUniqueID;
	$('.btn').attr('disabled',true);
	$('#vlnPlot').html('');
	$('#densityPlot').html('');
	$('#heatmapPlot').html('');
	n=3;

	$.ajax({
			url:url,
			dataType:'json',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
				n-=1;
				if(n==0)$('#loader').hide();
			},
			success:function(response){
				$('.btn').attr('disabled',false);
				$('#vlnPlot').html('<label style="display:block;text-align: center;">Violin Plot for target genes</label><br><br><img class="imgvln" src="data:image/svg+xml;base64,'+response.fileName+'" width="800" height="auto"/>');
				//wheelzoom(document.querySelectorAll('img.imgvln'));
			},
			error:function(response){
				alert(response.responseText);
				$('.btn').attr('disabled',false);
				return;
			}
	});
	url='/dgea/plot/?type=density&geneList='+tempValue;
	url+='&cID='+clientUniqueID;
	$('.btn').attr('disabled',true);
	$.ajax({
			url:url,
			dataType:'json',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
				n-=1;
				if(n==0)$('#loader').hide();
			},
			success:function(response){
				$('.btn').attr('disabled',false);
				$('#densityPlot').html('<label style="display:block;text-align: center;">Feature Plot for target genes</label><br><br><img class="feplot" src="data:image/svg+xml;base64,'+response.fileName+'" width="1000" height="auto"/>');
				//wheelzoom(document.querySelectorAll('img.feplot'));
			},
			error:function(response){
				alert(response.responseText);
				$('.btn').attr('disabled',false);
				return;
			}
	});
	url='/dgea/plot/?type=heatmap&geneList='+tempValue+'&groupby='+groupby;
	url+='&cID='+clientUniqueID;
	$('.btn').attr('disabled',true);
	$.ajax({
			url:url,
			dataType:'json',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
				n-=1;
				if(n==0)$('#loader').hide();
			},
			success:function(response){
				$('.btn').attr('disabled',false);
				$('#heatmapPlot').html('<label style="display:block;text-align: center;">Heatmap Plot for target genes(Scaled)</label><br><br><img class="heatplot" src="data:image/svg+xml;base64,'+response.fileName+'" width="1000" height="auto"/>');
				//wheelzoom(document.querySelectorAll('img.heatplot'));
			},
			error:function(response){
				alert(response.responseText);
				$('.btn').attr('disabled',false);
				return;
			}
	});
}

function LookUpGeneID()
{
	tempValue=$('#targetGene').val();
	sep=getSep();
	tempValue=tempValue.split(sep).join(',');
	if(tempValue=='')return;
	url='/GeneLookup/?geneList='+tempValue;
	url+='&cID='+clientUniqueID;
	$('.btn').attr('disabled',true);
	$.ajax({
			url:url,
			dataType:'text',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('.btn').attr('disabled',false);
				$('#targetGene').val(response.join(sep));
			},
			error:function(response){
				alert(response.responseText);
				$('.btn').attr('disabled',false);
				return;
			}
	});
}

//label_selected_value='';
function metaAdvanced()
{
	txt=$('#metaAdvancedSettingClick').html();
	flag=0;
	$('.btn').attr('disabled',true);
	if(integrated==0&&txt=='&gt;&gt;Enable Meta File Advanced Settings'){
		url='/user/?cID='+clientUniqueID;
		$.ajax({
			url:url,
			dataType:'text',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			success:function(response){
				$('.btn').attr('disabled',false);
			},
			error:function(response){
				$('.btn').attr('disabled',false);
				alert('Please Integrate/Process the data first.');
				flag=1;
			}
		});
		
	}
	if(flag==1)return;
	if(txt=='&gt;&gt;Enable Meta File Advanced Settings')
	{
		$('#metaAdvancedSettingClick').html('&gt;&gt;Close Meta File Advanced Settings');
		$('#metaAdvancedSetting').slideToggle(300);
		flag_open=1;
	}
	else
	{
		$('#metaAdvancedSettingClick').html('&gt;&gt;Enable Meta File Advanced Settings');
		$('#metaAdvancedSetting').slideToggle(300);
		flag_open=0;
		$('.btn').attr('disabled',false);
		return;
	}
	refreshMySelectLabel();
	
}

function refreshMySelectLabel()
{
	url='/meta/columns/?';
	url+='cID='+clientUniqueID;
	$.ajax({
			url:url,
			dataType:'json',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			$('.btn').attr('disabled',true);
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('.btn').attr('disabled',false);
				$('#my-select-label').html('');
				$('#my-select-label-numeric').html("<option value='' selected='selected'>----SELECT----</option>");
				for (i=0;i<response.length;i++)
				{
					colName=response[i][0];
					labeled=response[i][1];
					numeric=response[i][2];
					if(numeric=='0'){
						if(labeled=='0'){
							if(colName!='ID_REF') $('#my-select-label').append('<option value="'+colName+'">'+colName+'</option>');
							else $('#my-select-label').append('<option value="'+colName+'" disabled="">'+colName+'</option>');

						}
						else{
							if(colName!='LABEL'){
								$('#my-select-label').append('<option value="'+colName+'" selected="">'+colName+'</option>');
							}
							else{
								$('#my-select-label').append('<option value="batch2" selected="" disabled="" index="0">'+colName+'</option>');
								
							}
						}
					}
					else{
						$('#my-select-label-numeric').append('<option value="'+colName+'">'+colName+'</option>');
					}
				}
				$('#my-select-label').multiSelect('refresh');
				label_selected_value=$('#my-select-label').val();
				
			},
			error:function(response){
				alert(response.responseText);
				$('.btn').attr('disabled',false);
				return;
			}
	});
}

function refreshMyDeleteLabel()
{
	url='/meta/columns/?';
	url+='cID='+clientUniqueID;
	$.ajax({
			url:url,
			dataType:'json',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			$('.btn').attr('disabled',true);
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('.btn').attr('disabled',false);
				$('#my-select-label-delete').html("<option value='' selected='selected'>----SELECT----</option>");
				for (i=0;i<response.length;i++)
				{
					colName=response[i][0];
					labeled=response[i][1];
					numeric=response[i][2];
					if(numeric=='0'&&labeled=='0'&&colName!='ID_REF'&&(colName.indexOf("__crted") != -1)){
						$('#my-select-label-delete').append('<option value="'+colName+'">'+colName+'</option>');
					}

				}	
			},
			error:function(response){
				alert(response.responseText);
				$('.btn').attr('disabled',false);
				return;
			}
	});
}
function labelAssignConfirm()
{
	if($('#my-select-label').val()==''&&label_selected_value==''){
		alert('Select at least one categorical field as a new label.');
		return;
	}
	if($('#my-select-label').val()==label_selected_value){
		alert('Didn\'t change anything.');
		return;
	}
	url='/meta/columns/?';
	url+='labels='+$('#my-select-label').val()+'&fr='+$('input[name="FR"]:checked').val();
	url+='&cID='+clientUniqueID;
	$.ajax({
			url:url,
			dataType:'json',
			headers: {'X-CSRFToken': $('input[name="csrfmiddlewaretoken"]').val() },
			cache:false,
			contentType:false,
			processData:false,
			type:'put',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			$('.btn').attr('disabled',true);
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('.btn').attr('disabled',false);
				alert('Update new label(s) successfully.');
				label_selected_value=$('#my-select-label').val();
				$('#eda-label-choose').val('batch2');
				refreshMyDeleteLabel();
				
			},
			error:function(response){
				alert(response.responseText);
				$('.btn').attr('disabled',false);
				return;
			}
	});
}

function numLabelChange()
{
	colName=$('#my-select-label-numeric').val();
	if(colName=='')return;
	url='/meta/'+colName+'/?';
	url+='cID='+clientUniqueID;
	$.ajax({
			url:url,
			dataType:'text',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			$('.btn').attr('disabled',true);
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('.btn').attr('disabled',false);
				$('#numLabelPlot').html('<img src="data:image/svg+xml;base64,'+response+'" width="auto" height="600px"/>');
				
			},
			error:function(response){
				alert(response.responseText);
				$('.btn').attr('disabled',false);
				return;
			}
	});
}

function labelNumericConfirm()
{
	inp=$('#label-numeric-input').val();
	text=$('#label-numeric-input-texts').val();
	colName=$('#my-select-label-numeric').val();
	if(colName==''){
		alert('Please choose a field.');
		return;
	}
	pattern = /^(\d+(\.\d+)?)(,\d+(\.\d+)?)*$/;
	if(!pattern.test(inp))
	{
		alert('Illegal input for thredhold(s). Correct format case: 3.5,4.75');
		return;
	}
	pattern =/^[^,]+(?:,[^,]+)*$/;
	if(!pattern.test(text))
	{
		alert('Illegal input for values for thredhold(s). Correct format case: A,B,C');
		return;
	}
	inp=inp.split(',');
	text=text.split(',');
	if(text.length!=inp.length+1)
	{
		alert('Two inputs don\'t match.');
		return;
	}
	url='/meta/columns/?';
	url+='cID='+clientUniqueID;
	$.ajax({
			url:url,
			dataType:'text',
			cache:false,
			data:JSON.stringify({
				'colName':colName,
				'thredshold':inp,
				'thredshold_labels':text
			}),
			headers: {'X-CSRFToken': $('input[name="csrfmiddlewaretoken"]').val() },
			contentType:'application/json',
			processData:false,
			type:'post',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			$('.btn').attr('disabled',true);
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('.btn').attr('disabled',false);
				alert('Label Created Successfully.');
				$('#numLabelPlot').html('');
				$('#label-numeric-input').val('');
				$('#label-numeric-input-texts').val('');
				$('#my-select-label-numeric').val('');
				refreshMySelectLabel();			
			},
			error:function(response){
				alert(response.responseText);
				$('.btn').attr('disabled',false);
				return;
			}
	});
}

function labelDeleteConfirm()
{
	value=$('#my-select-label-delete').val();
	if(value=='')return
	url='/meta/'+value+'/';
	url+='?cID='+clientUniqueID;
	$.ajax({
			url:url,
			dataType:'text',
			cache:false,
			headers: {'X-CSRFToken': $('input[name="csrfmiddlewaretoken"]').val() },
			contentType:'application/json',
			processData:false,
			type:'delete',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			$('.btn').attr('disabled',true);
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('.btn').attr('disabled',false);
				alert('Label Deleted Successfully.');
				refreshMySelectLabel();
				refreshMyDeleteLabel();		
			},
			error:function(response){
				alert(response.responseText);
				$('.btn').attr('disabled',false);
				return;
			}
	});
}

function edaLabelChange()
{
	$('#next2EDA').trigger('click');
}

function label_name_change(targetLabel,targetClusterNum)
{
	if($('#'+targetLabel).val()=='')return;
	url='/meta/'+$('#'+targetLabel).val()+'/';
	url+='?cID='+clientUniqueID;
	$.ajax({
			url:url,
			dataType:'json',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			success:function(response){
				$('.btn').attr('disabled',false);
				$('#'+targetClusterNum).html("<option value='' selected='selected'>----SELECT----</option>");
				for (i=0;i<response.length;i++){
					$('#'+targetClusterNum).append(new Option(response[i],response[i]));
				}
				
			},
			error:function(response){
				alert(response.responseText);
				$('.btn').attr('disabled',false);
				return;
			}
	});
}

  </script>
  
  <style>
  #loader{
  position:fixed;
  left:0px;
  top:0px;
  width:100%;
  height:100%;
  z-index:99999;
  background:url(/static/loading12.gif) 50% 50% no-repeat rgb(15 10 10 /59%);
  background-size:200px;
  }
  .rotated{
  transform:rotate(90deg);
  }

  </style>
  
<div id='loader' style='display:none'></div>
<div id="tabs" style='width:1800px;'>
  <ul>
    <li><a href="#tabs-1" class="first">Data Collection</a></li>
    <li><a href="#tabs-2" class="second">Data Processing</a></li>
    <li><a href="#tabs-3" class="third">EDA</a></li>
    <li><a href="#tabs-4" class="fourth">DGEA-Clustering</a></li>
	<li><a href='#tabs-7' class="fifth">DGEA-Target Gene</a></li>
    <li><a href="#tabs-5" class="sixth">LASSO FeastureSelect</a></li>
    <li><a href="#tabs-6" class="seventh">Go Enrichment</a></li>
  </ul>
  <div id="tabs-1">
    <fieldset name="Multiple Files Upload">
    <div id="msg"></div>
    <p>Upload Gene Expression files&nbsp;<img id='tooltip1'class='question' title="In this section, users upload their files(multiple). File names will be used for later visualisation. ID_REF is a compulsory field for each file.<p><img src='/static/fileUpload.png' width='500' height='auto'/></p>"/>&nbsp;&nbsp;&nbsp;
		<button class="btn btn-primary" id='TemplateUpload' onclick="window.location.href='/static/template.zip'">Download Templates for Upload</button> &nbsp;&nbsp;&nbsp;
		<button id="advancedSearch" class="btn btn-primary" onclick="window.open('/advancedSearch/?cID='+clientUniqueID);" title="This is used for processing any online data. User can search for any public dataset and integrate them into the research.">Advanced Search</button></p>
    <table>
    <tr>
	<td style="vertical-align:top;">
	{% csrf_token %}
		<input type="file" id="multiFiles" name="files[]" multiple="multiple" title="Choose from local"/>
		<button id="upload1" class="btn btn-primary">Upload</button>
	</td>

    </tr>
    <tr>
  	<td>
  	<div id='preview1' style="height:400px; width:1200px; overflow-x:scroll; overflow-y:scroll;"></div>	
  	</td>
    </tr>
    </table>
    
    <p>Upload Meta-data file&nbsp;<img class='question' title="In this section, users upload their meta-data file(single). ID_REF and LABEL are compulsory. ID_REF is matched with the previous ones in Expression files. LABEL will be used for later visualisation. User can upload clinic data as well.<p><img src='/static/metaUpload.png' width='auto' height='auto' /></p>"/></p>
    <table>
    <tr>
	<td style="vertical-align:top;">
	{% csrf_token %}
		<input type="file" id="singleFile" name="singleFile" title="Choose from local"/>
		<button id="upload2" class="btn btn-primary">Upload</button>
		<button id="next1" class="btn btn-primary">&nbsp;&nbsp;Next&nbsp;&nbsp;</button>	
	</td>

    </tr>
    <tr>
  	<td>
  	<div id='preview2' style="height:400px; width:1200px; overflow-x:scroll; overflow-y:scroll;"></div>	
  	</td>
    </tr>
    </table>
   </fieldset>
  </div>
  
  <div id="tabs-2">
  <table>
  
  <tr>
	<div id="built-in-data">
  	<h5>Choose the built-in compiled cohort(s) you want to integrate with:</h5>
	<br><br>
	<select id="my-select" multiple="multiple">
	<optgroup label='Shared-Cohorts'>
		{% for cohort in root.cohorts %}
    		<option value="{{ cohort.0 }}" title="{{ cohort.1 }}">{{ cohort.0 }}</option>
    	{% endfor %}
    	</optgroup>
	</select>
	</div>
  </tr>

  	
  <tr id="processOption">
  	<td>
   	<div name="Options1" style="left:30px; width:300px;">
   		<br>
		<p>Combined Options</p>
		<input type="radio" name="Comb" value="Combat" checked="checked" /><label>&nbsp;Combat&nbsp;</label><img class='question' style="max-width:4%;" title="ComBat, originally implemented in the R library sva. This tool leverages a parametric and non-parametric empirical Bayes approach for correcting the batch effect in microarray datasets that works for small sample sizes or in the presence of outliers."/>
		<br>
  		<input type="radio" name="Comb" value="Harmony" /><label>&nbsp;Harmony&nbsp;</label><img class='question' style="max-width:4%;" title="<p>Harmony is a general-purpose package with an efficient algorithm for integrating multiple data sets. It is especially useful for large single-cell datasets such as single-cell RNA-seq.</p><p>Harmony is:</p><p><b>Fast:</b> Analyze thousands of cells on your laptop.</p><p><b>Sensitive:</b> Different cell types may be present or absent in each batch.</p><p><b>Accurate:</b> Integrate cells from multiple donors, tissues – even different technologies.</p>"/>
  		<br>
  		<br>
	</div>
	</td>
	
	<td>
	<div name="Options2" style="left:30px; width:300px;">
		<br>
		<p>Log2 Transformation</p>
		<input type="radio" name="Log" value="Yes" /><label>&nbsp;Yes</label>
		<br>
  		<input type="radio" name="Log" value="No" checked="checked"><label>&nbsp;No</label>
		<br>
		<br>
	</div>
	</td>

	<td>
	<div name="Options3" style="left:30px; width:300px;">
		<br>
		<p>Feature Reduction Method</p>
		<input type="radio" name="FR" value="TSNE" /><label>&nbsp;TSNE&nbsp;</label><img class='question' style="max-width:4%;" title="<p>t-SNE (t-distributed Stochastic Neighbor Embedding) is an unsupervised non-linear dimensionality reduction technique for data exploration and visualizing high-dimensional data. Non-linear dimensionality reduction means that the algorithm allows us to separate data that cannot be separated by a straight line. </p>"/>
		<br>
  		<input type="radio" name="FR" value="UMAP"  checked="checked" /><label>&nbsp;UMAP&nbsp;</label><img class='question' style="max-width:4%;" title="<p>Uniform Manifold Approximation and Projection (UMAP) is a dimension reduction technique that can be used for visualisation similarly to t-SNE, but also for general non-linear dimension reduction.</p>"/>
		<br>
		<input type="radio" name="FR" value="PCA" /><label>&nbsp;PCA&nbsp;</label><img class='question' style="max-width:4%;" title="<p>Principal Component Analysis</p>"/>
		<br>
		<br>
	</div>
	</td>
	<td>
  	</td>
	</td>
  </tr>
  <tr>
	<td>
		<div id="metaAdvancedSettingClick" onclick="metaAdvanced();" title="Click to create new labels based on Meta File Data.">&gt;&gt;Enable Meta File Advanced Settings</div>
	</td>
  	<td style="text-align:right;" colspan='2'>
		<button id="next2" class="btn btn-danger">&nbsp;&nbsp;Processing&nbsp;&nbsp;</button>
		<button id="next2EDA" class="btn btn-primary" disabled>&nbsp;&nbsp;Next&nbsp;&nbsp;</button>
	</td>
  </tr>
  </table>
  <div id="metaAdvancedSetting">
	
	<div id="tabs-label" style='width:1250px;'>
		<ul>
		  <li><a href="#tabs-label-1">Label Generation</a></li>
		  <li><a href="#tabs-label-2">Label Assignment</a></li>
		  <li><a href="#tabs-label-3" onClick="refreshMyDeleteLabel();">Label Delete</a></li>
		</ul>
		<div id="tabs-label-1">
			<label>Create New Label from numerical fields:</label>
			<img class='question' style="max-width:0.8%;" title="This tab is used for creating labels based on Continuous variables in meta file.<br>After you created the label, you will need to go to the next tab to activate it.<img src='/static/continuous.png' width='500' height='auto'/>"/>
			<br>
			<select id="my-select-label-numeric" onchange="numLabelChange();">
				<option value='' selected='selected'>----SELECT----</option>
			</select>
			<br><br><br>
			<div>
				<ul>
					<li>
						Input the numeric thredshold(s) and separate each by Comma.
						<br>
						<input id="label-numeric-input" style="width:600px"/>
					</li>
					<br>
					<li>
						Input the Labels corresponding to the intervals and separate each by Comma.
						<br>
						<input id="label-numeric-input-texts" style="width:600px"/>
						&nbsp;&nbsp;<button id="label-numeric-confirm" class="btn btn-primary" onclick="labelNumericConfirm();">&nbsp;&nbsp;Add New Label&nbsp;&nbsp;</button>
					</li>
				</ul>
			</div>
			<br><br>
			<div id='numLabelPlot' style="width:auto; height:600px; float:left;"></div>
		</div>
		<div id="tabs-label-2">
			<div style="text-align: center;">
				<div style="display: inline-block;">
					<br><br>
					<select id="my-select-label" multiple="multiple">
					</select>
					<br>
				</div>
			</div>
			<ul>
				<li>Press 'Modify' to make the above selected categorical fields as Labels:<img class='question' style="max-width:0.8%;" title="Activate you previously created label or the Categorical label in the meta-file.<br><img src='/static/labels.png' width='500' height='auto'/>"/>
				</li>
			</ul>
			<button id="label-assign-confirm" class="btn btn-primary" onclick="labelAssignConfirm();"; style="margin-left:800px;" style="margin-left:300px;">&nbsp;&nbsp;Modify&nbsp;&nbsp;</button>	
		</div>
		<div id="tabs-label-3">
			<ul>
				<li>Delete the label you previously created.<img class='question' style="max-width:0.8%;" title="Only inactive labels with the suffix  '__crted'  can be deleted."/>
				</li>
			</ul>
			<div style="text-align: center;">
				<div style="display: inline-block;">
					<br><br>
					<select id="my-select-label-delete">
						<option value='' selected='selected'>----SELECT----</option>
					</select>
					<br>
				</div>
			</div>
			<button id="label-delete-confirm" class="btn btn-primary" onclick="labelDeleteConfirm();"; style="margin-left:800px;" style="margin-left:300px;">&nbsp;&nbsp;Delete&nbsp;&nbsp;</button>	
		
		</div>
	</div>
  </div>
  </div>
  

  <div id="tabs-3">
  	<p>Gene Data Explory Data Analyses</p>
	<br>
	<div  id='VbOption' style="margin: 0 auto; display: table; background-color: rgb(184, 184, 224); border-radius: 5px;">
		<label>&nbsp;&nbsp;Choose the created Label as target:&nbsp;&nbsp;</label>
		<select id="eda-label-choose" onchange="edaLabelChange();">
			<option value='batch2' selected='selected'>LABEL</option>
		</select>&nbsp;&nbsp;
	</div>
  	<div id='tab3div'>
		<div id="tabs-eda" style='width:1250px;'>
			<ul>
			  <li><a href="#tabs-eda-1" id="VbL">Data processed by Labels</a></li>
			  <li><a href="#tabs-eda-2" id="VbF">Data processed by FileNames</a></li>
			</ul>
			<div id="tabs-eda-1">
				<div id="graphDiv2" class='no-break' style="width:850px;height:850px;overflow-x:scroll; overflow-y:scroll;"></div>
				<p>Dotplot of top 4 genes for each LABEL</p>
				<a id='graphDiv21A' href="..." target="_blank" download="image.svg" style="display: block;">Download Image</a><br><br><br><br><br><br><br><br><br>
  				<div id="graphDiv21" class="rotated no-break" style="width:800px;height:500px; overflow-x:scroll; overflow-y:scroll;"></div>
				<br><br><br><br><br><br><br><br><br>
  				<button id='topForLabelsDownloadBtn' class='btn btn-primary' onclick='getTopGenes(2);'>Download DEGs</button>		
			</div>
			<div id="tabs-eda-2">
				<div id="graphDiv1" class='no-break' style="width:850px;height:850px;overflow-x:scroll; overflow-y:scroll;"></div>
				<div style="width:100%; border-top:1px solid #ccc;"></div>
				<p>Dotplot of top 4 genes for each file</p>
				<a id='graphDiv11A' href="..." target="_blank" download="image.svg" style="display: block;">Download Image</a><br><br><br><br><br><br><br><br><br>
				<div id="graphDiv11" class="rotated no-break" style="width:800px;height:500px; overflow-x:scroll; overflow-y:scroll;"></div>
				<br><br><br><br><br><br><br><br><br>
				<button id='topForFilesDownloadBtn' class='btn btn-primary' onclick='getTopGenes(1);'>Download DEGs</button>
			</div>
		</div>
  	</div>
  	<br>
	<div style="text-align:right;">
		<button id="download1" class="btn btn-primary" style="width:auto;" onclick='getTopGenes(4);'>Download Processed data</button>
		<br><br>
		<button id="next3" class="btn btn-primary">&nbsp;&nbsp;Next&nbsp;&nbsp;</button>
			
	</div>

  </div>
 
 
   <div id="tabs-4">
  	<p>Choose Options for Clustering:&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
  	
	<table>
			<tr>
				<td style="height:50px; width:200px; border-right:1px solid #ccc;">
					<input type="radio" name='clustering' value='LEIDEN' checked /><label>&nbsp;LEIDEN&nbsp;</label><img class='question' style="max-width:6.5%;" title="<p>The Leiden algorithm is an algorithm for detecting communities in large networks. The algorithm separates nodes into disjoint communities so as to maximize a modularity score for each community. Modularity quantifies the quality of an assignment of nodes to communities, that is how densely connected nodes in a community are, compared to how connected they would be in a random network.<p></p>The Leiden algorithm is a hierarchical clustering algorithm, that recursively merges communities into single nodes by greedily optimizing the modularity and the process repeats in the condensed graph. It modifies the Louvain algorithm to address some of its shortcomings, namely the case where some of the communities found by Louvain are not well-connected. This is achieved by periodically randomly breaking down communities into smaller well-connected ones.</p>"/>
					<br>
					<input type="radio" name='clustering' value='HDBSCAN' /><label>&nbsp;HDBSCAN&nbsp;</label><img class='question' style="max-width:6.5%;" title="<p>HDBSCAN is a clustering algorithm developed by Campello, Moulavi, and Sander. It extends DBSCAN by converting it into a hierarchical clustering algorithm, and then using a technique to extract a flat clustering based in the stability of clusters. The goal of this notebook is to give you an overview of how the algorithm works and the motivations behind it. In contrast to the HDBSCAN paper I’m going to describe it without reference to DBSCAN. Instead I’m going to explain how I like to think about the algorithm, which aligns more closely with Robust Single Linkage with flat cluster extraction on top of it.</p>"/>
					<br>
					<input type="radio" name='clustering' value='KMeans' /></label>&nbsp;KMeans&nbsp;</label><img class='question' style="max-width:6.5%;" title="<p>k-means clustering is a method of vector quantization, originally from signal processing, that aims to partition n observations into k clusters in which each observation belongs to the cluster with the nearest mean (cluster centers or cluster centroid), serving as a prototype of the cluster. This results in a partitioning of the data space into Voronoi cells. k-means clustering minimizes within-cluster variances (squared Euclidean distances), but not regular Euclidean distances, which would be the more difficult Weber problem: the mean optimizes squared errors, whereas only the geometric median minimizes Euclidean distances. For instance, better Euclidean solutions can be found using k-medians and k-medoids.</p>"/>
				</td>	
				<td style="width:400px; height:50px; border-right:1px solid #ccc;">
					<div id='LPara'>
						<label for='slider1'>Input Param: Resolution  <img class='question' style="max-width:3%;" title="A parameter value controlling the coarseness of the clustering. Higher values lead to more clusters."/> = </label>
						<input type='text' id='sliderValue' size='6'></input>
						<br><br>
						<div id='slider1'></div>
					</div>
					<div id='HPara'>
						<label for='slider1'>Input Param: minSize  <img class='question' style="max-width:3%;" title="The minimum number of samples in a group for that group to be considered a cluster; groupings smaller than this size will be left as noise."/> = &nbsp;</label>
						<br>
						<input id='hdbP'  value='5' />
					</div>
					<div id='KPara'>
						<label for='slider1'>Input Param: K  <img class='question' style="max-width:3%;" title="The number of clusters to form as well as the number of centroids to generate."/> =  &nbsp; &nbsp;</label>
						<br>
						<input id='kP'  value='5' />
					</div>
				</td>
				<td style="height:50px;">
				<button id="clusterBtn" class="btn btn-primary" style="width:auto;" onclick="clusterBtnClick();">Cluster the Data</button>
				<button id='advanced' class="btn btn-primary" style='width:auto;' onclick="window.open('/cluster/advanced/?cID='+clientUniqueID);">Advanced</button>
				</td>
			</tr>
	</table>
	<div id='tab4div'>
		<div id="tabs-dgea" style='width:1250px;'>
			<ul>
			  <li><a href="#tabs-dgea-1" id="VbC">Data colored by clusters</a></li>
			  <li><a href="#tabs-dgea-2" id="DE4C">Dotplot of top 4 genes for each cluster</a></li>
			</ul>
			<div id="tabs-dgea-1">
				<div id='graphDiv1-cluster' class='no-break' style="width:850px; height:850px; overflow-x:scroll; overflow-y:scroll;"></div>
			</div>
			<div id="tabs-dgea-2">
				<br><br><br>
					<a id="graphDiv1-cluster-gene-A" href="..."  target="_blank" download="image.svg" style="display: block;">Download Image</a><br><br>
              		<div id="graphDiv1-cluster-gene" class='rotated no-break' style="width:600px;height:500px; overflow-x:scroll; overflow-y:scroll;"></div>
					<br><br><br>
					<button id='topForLabelsDownloadBtn' class='btn btn-primary' onclick='getTopGenes(3);'>Download DEGs</button>
			</div>
		</div>
	<table>
			<tr>
				<td><p>Contributions to the clusters from each file</p></td>
				<td><p>Contributions to the clusters from each LABEL</p></td>
			</tr>
			<tr>
				<td>
				<div id='bc1' class='no-break' style="width:600px; height:500px; overflow-x:scroll; overflow-y:scroll;"></div>
				</td>
				<td>	
				<div id="bc2" class='no-break' style="width:600px; height:500px;overflow-x:scroll; overflow-y:scroll;"></div>
				</td>
			</tr>
			<tr>
				<td colspan='2'>
				<p>Top 20 genes after running DE for clusters</p>
				<a id="graphDiv1-cluster-top-gene-A" href="..."  target="_blank" download="image.svg" style="display: block;">Download Image</a><br><br>
				<div id='graphDiv1-cluster-top-gene' class='no-break' style="width:1000px; overflow-x:scroll; overflow-y:scroll;"></div>
				</td>
			</tr>
			<tr class='downloadData1'>
				<td style="text-align:right;" colspan='3'>
				<button id="download2" class="btn btn-primary" style="width:auto;" onclick='getTopGenes(5);'>Download Processed data</button>
				<br><br>
					<button id="next4" class="btn btn-primary">&nbsp;&nbsp;Next&nbsp;&nbsp;</button>
				</td>
			</tr>
	</table>
	</div>	
	</table>
  </div>

  <div id="tabs-7">
	<p>Input your target gene list for the groups for Violin, density and heatmap plot and choose the separator splitting the genes.
		<img class='question' style="max-width:0.8%;" title="This tab is used for plots for at most 15 target Genes after running clustering.<br>You can input your interested genes, generating its violin/density and heatmap plots. <br>If your input is GeneID, you may need to convert them to Symbol ID by clicking the 'Convert to Symbol ID' button first.<br>Sys provides you candidate gene Lists, <ul><li>PCA(3) for top important genes contributing to the first 3 Principal Components after PCA.</li><li>The second option is for marker genes after clustering.</li></ul>"/>	
	</p>
	<label>Choose the separator for the gene list:</label>
	<select id="separator">
		<option value="Comma"  selected="selected">Comma</option>
		<option value="Return">Return</option>
		<option value="Space">Space</option>	
		<option value="Tab">Tab</option>
		<option value="Semicolon">Semicolon</option>
	</select>&nbsp;&nbsp;&nbsp;
	<label>Select Cluster:</label>
	<select id="candidateCluster" onchange='$("#candidateGeneList option:first").prop("selected", true);'>
		<option value="" selected="selected"></option>
		<option value="cluster">Clustering Method</option>
	</select>
	<label>Choose Candidate Gene List:</label>
	<select id="candidateGeneList" onchange="loadCandiGenes();">
		<option value="" selected="selected"></option>
		<option value="pca">PCA(3)</option>
		<option value="clusters">Related to Cluster</option>
	</select>
	<br><br>
	<div style="position:relative;display:inline-block;">
	<textarea id="targetGene"  rows='5' cols='100'></textarea>
	&nbsp;&nbsp;<button onclick='LookUpGeneID();' class="btn btn-primary" style="position:absolute; width:120px; height:120px;top:15px;">Convert to Symbol ID</button>
	</div>
	<br>
	<button onclick='TGenePlotClick();' class="btn btn-primary" style="position: relative;left:950px;">Submit</button>
	<br><br>
	<div id="vlnPlot" class='no-break'></div>
	<br><br>
	<div id="densityPlot" class='no-break'></div>
	<br><br>
	<div id="heatmapPlot" class='no-break'></div>

  </div>
  
  <div id='tabs-5'>
  	<p>Select one Label/Cluster of interest for Feature Selection by Lasso:&nbsp;<img class='question' style="max-width:0.8%;" title="<p>LASSO regression, also known as L1 regularization, is a popular technique used in statistical modeling and machine learning to estimate the relationships between variables and make predictions. LASSO stands for Least Absolute Shrinkage and Selection Operator.</p><p>The primary goal of LASSO regression is to find a balance between model simplicity and accuracy. It achieves this by adding a penalty term to the traditional linear regression model, which encourages sparse solutions where some coefficients are forced to be exactly zero. This feature makes LASSO particularly useful for feature selection, as it can automatically identify and discard irrelevant or redundant variables.</p>"/></p>
	<table>
  		<tr>
			<td style="text-align:left;width:50%">
				<label>Select Label/Cluster:</label>
				<br>
				<select id='lasso_label_name' onchange="label_name_change('lasso_label_name','cluster_num1');">
					<option value='' selected='selected'>----SELECT----</option>
				</select>
				<br><br>
				<label>Select group of interest of the Label/Cluster:</label>
				<br>
				<select id='cluster_num1'>
					<option value='' selected='selected'>----SELECT----</option>
				</select>
			</td>
			<td style="text-align:left;width:50%">
				<button id="cluster_btn" class="btn btn-primary" onclick="golasso();">Lasso</button>
			</td>
		</tr>
		<tr>
		<td colspan='2'>
		<div id='lassoPic' class='no-break' style="width:1200px; height:800px; overflow-x:scroll; overflow-y:scroll; text-align:center;"></div>
		</td>
		</tr>
  	</table>
  </div>
  
  <div id='tabs-6'>
  	<p>Select one Label/Cluster of interest for Go Enrichment Analysis:&nbsp;<img class='question' style="max-width:0.8%;" title="<p>GO (Gene Ontology) enrichment analysis is a computational method used in bioinformatics to identify overrepresented gene ontology terms associated with a set of genes. Gene Ontology provides a standardized vocabulary for annotating the molecular functions, biological processes, and cellular components of genes and gene products.</p>"/></p>
  	<table>
  		<tr>
			<td style="text-align:left;width:50%">
				<label>Select Label/Cluster:</label>
				<br>
				<select id='go_label_name' onchange="label_name_change('go_label_name','cluster_num');">
					<option value='' selected='selected'>----SELECT----</option>
				</select>
				<br><br>
				<label>Select Class of interest of Cluster:</label>
				<br>
				<select id='cluster_num'>
					<option value='' selected='selected'>----SELECT----</option>
				</select>
			</td>
			<td style="text-align:left;width:50%;">
			<button id="cluster_btn" class="btn btn-primary" onclick="goEnrich();">Go Enrichment</button>
			</td>
		</tr>
		<tr>
		<td colspan='2'>
		<div id='goEnrichPic' class='no-break' style="width:1200px; overflow-x:scroll; overflow-y:scroll; text-align:center;"></div>
		</td>
		</tr>
  	</table>
  </div>

{% endblock %}
